src/Function/Bundles/Related.agda

```diff
@@ -0,0 +1,399 @@
+equivalence         op = equivalence
+injection           op = reverse-injection
+reverse-injection   op = injection
+left-inverse        op = surjection
+surjection          op = left-inverse
+bijection           op = bijection
+
+-- For every morphism there is a corresponding reverse morphism of the
+-- opposite kind.
+
+reverse : ∀ {k a b} {A : Set a} {B : Set b} →
+          A ∼[ k ] B → B ∼[ k op ] A
+reverse {implication}         = id
+reverse {reverse-implication} = id
+reverse {equivalence}         = Symmetry.sym-⇔
+reverse {injection}           = id
+reverse {reverse-injection}   = id
+reverse {left-inverse}        = ↪⇒↠
+reverse {surjection}          = ↠⇒↪
+reverse {bijection}           = ↔⇒⤖ ∘ Symmetry.sym-↔ ∘ ⤖⇒↔
+
+------------------------------------------------------------------------
+-- For a fixed universe level every kind is a preorder and each
+-- symmetric kind is an equivalence
+
+K-refl : ∀ {k ℓ} → Reflexive (Related k {ℓ})
+K-refl {implication}         = Identity.id-⟶ _
+K-refl {reverse-implication} = Identity.id-⟶ _
+K-refl {equivalence}         = Identity.id-⇔ _
+K-refl {injection}           = Identity.id-↣ _
+K-refl {reverse-injection}   = Identity.id-↣ _
+K-refl {left-inverse}        = Identity.id-↪ _
+K-refl {surjection}          = Identity.id-↠ _
+K-refl {bijection}           = Identity.id-⤖ _
+
+K-reflexive : ∀ {k ℓ} → P._≡_ Relation.Binary.⇒ Related k {ℓ}
+K-reflexive P.refl = K-refl
+
+K-trans : ∀ {k ℓ₁ ℓ₂ ℓ₃} → Trans (Related k {ℓ₁} {ℓ₂})
+                                 (Related k {ℓ₂} {ℓ₃})
+                                 (Related k {ℓ₁} {ℓ₃})
+K-trans {implication}         = Composition._∘-⟶_
+K-trans {reverse-implication} = flip Composition._∘-⟶_
+K-trans {equivalence}         = Composition._∘-⇔_
+K-trans {injection}           = Composition._∘-↣_
+K-trans {reverse-injection}   = flip Composition._∘-↣_
+K-trans {left-inverse}        = Composition._∘-↪_
+K-trans {surjection}          = Composition._∘-↠_
+K-trans {bijection}           = Composition._∘-⤖_
+
+SK-sym : ∀ {k ℓ₁ ℓ₂} → Sym (Related ⌊ k ⌋ {ℓ₁} {ℓ₂})
+                          (Related ⌊ k ⌋ {ℓ₂} {ℓ₁})
+SK-sym {equivalence} = reverse
+SK-sym {bijection}   = reverse
+
+SK-isEquivalence : ∀ k ℓ → IsEquivalence {ℓ = ℓ} (Related ⌊ k ⌋)
+SK-isEquivalence k ℓ = record
+  { refl  = K-refl
+  ; sym   = SK-sym
+  ; trans = K-trans
+  }
+
+SK-setoid : Symmetric-kind → (ℓ : Level) → Setoid _ _
+SK-setoid k ℓ = record { isEquivalence = SK-isEquivalence k ℓ }
+
+K-isPreorder : ∀ k ℓ → IsPreorder _⤖_ (Related k)
+K-isPreorder k ℓ = record
+    { isEquivalence = SK-isEquivalence bijection ℓ
+    ; reflexive     = ⤖⇒
+    ; trans         = K-trans
+    }
+
+K-preorder : Kind → (ℓ : Level) → Preorder _ _ _
+K-preorder k ℓ = record { isPreorder = K-isPreorder k ℓ }
+
+------------------------------------------------------------------------
+-- Equational reasoning
+
+-- Equational reasoning for related things.
+
+module EquationalReasoning where
```

Comment:
The other reasoning APIs have come a long way since this one was created. We can probably use `Relation.Binary.Reasoning.Base.Double` as a base module for this? And should update the style accordingly.

Reply:
Because one can reason with inverses and bijections, `Triple` seems appropriate. However, these modules all seem to have the interpretation that the extra relations are not symmetric. I don't know if this would cause any trouble though.

Reply:
Actually, it seems to me like the most natural form of this would be some form of `n-ary` reasoning, with a bunch of equivalence relation and orderings and implications between them. I'm not sure if that's possible though.